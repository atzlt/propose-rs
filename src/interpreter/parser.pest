main = {
    SOI ~ file_line* ~ eoi
}

file_line = { (draw | decor | decl | config_line | save_file) ~ ";" }

decl       = { decl_left ~ "=" ~ decl_right }
decl_left  = { destruct | direct }
direct     = { any_id }
destruct   = { any_id ~ "," ~ any_id }
decl_right = { coord | expr | arg | eval }

expr   =  { method ~ args }
method = @{ punc_alpha+ }
args   =  { (arg ~ ",")* ~ arg }

arg = {
    trig
  | common_obj
  | numeric
  | eval
}

draw       =  { "draw" ~ (draw_step ~ ",")* ~ draw_step }
draw_step  =  { draw_obj ~ ("[" ~ configs ~ "]")? }
draw_obj   =  {
    polygon
  | arc
  | arc_o
  | common_obj
}
decor      = { "decor" ~ (decor_step ~ ",")* ~ decor_step }
decor_step = { decor_obj ~ ":" ~ decoration ~ ("[" ~ configs ~ "]")? }
decor_obj  = {
    polygon
  | angle_3p
  | arc
  | arc_o
  | common_obj
}
decoration = @{ punc_alpha+ }

config_line  =  { "config" ~ configs }
configs      =  { (config ~ ",")* ~ config }
config       =  { config_name ~ "=" ~ config_value }
config_name  = @{ ASCII_ALPHA_LOWER+ }
config_value =  { rich_number | boolean | string }

save_file = { "save" ~ raw_string }

trig       = ${ point_id ~ point_id ~ point_id }
line_2p    =  { point_id ~ point_id }
circ_3p    =  { "@(" ~ point_id ~ point_id ~ point_id ~ ")" }
circ_or    =  { "@(" ~ point_id ~ "," ~ numeric ~ ")" }
circ_oa    =  { "@(" ~ point_id ~ "," ~ point_id ~ ")" }
circ_diam  =  { "@(" ~ point_id ~ point_id ~ ")" }
polygon    = ${ point_id ~ "-" ~ (point_id ~ "-")+ ~ point_id }
arc        = ${ point_id ~ "~" ~ point_id ~ "~" ~ point_id }
arc_o      = ${ point_id ~ "~(" ~ point_id ~ ")~" ~ point_id }
common_obj =  {
    line_2p
  | circ_3p
  | circ_or
  | circ_oa
  | circ_diam
  | any_id
}

linear      = { line_2p | common_id }
distance    = { ("|" ~ point_id ~ point_id ~ "|") | ("|(" ~ (linear | point_id) ~ "," ~ linear ~ ")|") }
angle_3p    = { "<" ~ point_id ~ point_id ~ point_id }
angle_2l    = { "<(" ~ linear ~ "," ~ linear ~ ")" }
numeric     = { distance | angle_3p | angle_2l | rich_number | common_id }

coord       = { "(" ~ (ortho_coord | polar_coord) ~ ")" }
ortho_coord = { numeric ~ "," ~ numeric }
polar_coord = { numeric ~ ":" ~ numeric }

eval = { "$" ~ (!"$" ~ ANY)* ~ "$" }

raw_string  = @{ (!(NEWLINE | ";") ~ ANY)* }
rich_number =  { degree | number }
boolean     =  { "true" | "false" }
degree      = @{ number ~ "deg" }
number      = @{ ("+" | "-")? ~ ASCII_DIGIT+ ~ ("." ~ ASCII_DIGIT+)? }
string      = ${ "\"" ~ str_inner ~ "\"" }
str_inner   = @{ char* }
char        =  {
    !("\"" | "\\") ~ ANY
  | "\\" ~ ("\"" | "\\" | "/" | "b" | "f" | "n" | "r" | "t")
  | "\\" ~ ("u" ~ ASCII_HEX_DIGIT{4})
}
punc_alpha  = _{
    ASCII_ALPHA
  | "#"
  | "@"
  | "*"
  | "<"
  | "-"
  | ">"
  | "!"
  | "?"
  | "/"
  | "\\"
  | "|"
}

point_id  = @{ ASCII_ALPHA_UPPER ~ (ASCII_ALPHA_LOWER | ASCII_DIGIT)* ~ "'"? }
common_id = @{ (ASCII_ALPHA_LOWER | "_") ~ (ASCII_ALPHA_LOWER | ASCII_DIGIT)* ~ "'"? }
any_id    =  { point_id | common_id }

eoi = _{ !ANY }
WHITESPACE = _{ (NEWLINE | " " | "\t")+ }
