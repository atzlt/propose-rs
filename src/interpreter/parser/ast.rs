use crate::interpreter::utils::ConfigValue;
#[cfg(test)]
use serde::Serialize;
use std::collections::HashMap;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum Linear {
    Line2P(String, String),
    Name(String),
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum Numeric {
    Distance2P(String, String),
    DistancePL(String, Linear),
    Distance2L(Linear, Linear),
    Angle3P(String, String, String),
    Angle2L(Linear, Linear),
    Number(f64),
    Name(String),
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum Object {
    Line2P(String, String),
    Circ3P(String, String, String),
    CircOr(String, Box<Numeric>),
    CircOA(String, String),
    CircDiam(String, String),
    Arc(String, String, String),
    ArcO(String, String, String),
    Angle3P(String, String, String),
    Triangle(String, String, String),
    Polygon(Vec<String>),
    Name(String),
    Numeric(Box<Numeric>),
    Eval(String),
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum DeclRight {
    OrthoCoord(Box<Numeric>, Box<Numeric>),
    PolarCoord(Box<Numeric>, Box<Numeric>),
    Expr(String, Vec<Object>),
    Object(Box<Object>),
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum DeclLeft {
    Direct(String),
    Destruct(String, String),
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub struct Decl(pub DeclLeft, pub DeclRight);

pub type Config = HashMap<String, ConfigValue>;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub struct StyledObject {
    pub obj: Object,
    pub config: Option<Config>,
}

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub struct DecorObject {
    pub obj: Object,
    pub decor: String,
    pub config: Option<Config>,
}

pub type Draw = Vec<StyledObject>;
pub type Decor = Vec<DecorObject>;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug)]
pub enum FileLine {
    Config(Config),
    Draw(Draw),
    Decor(Decor),
    Decl(Box<Decl>),
}

pub type Main = Vec<FileLine>;
