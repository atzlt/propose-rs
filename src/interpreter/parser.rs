pub mod ast;

use super::utils::ConfigValue;
use ast::*;
use pest_consume::Parser;
use pest_consume::{match_nodes, Error};
use std::f64::consts::PI;

type Result<T> = std::result::Result<T, Error<Rule>>;
type Node<'i> = pest_consume::Node<'i, Rule, ()>;

#[derive(Parser)]
#[grammar = "interpreter/parser.pest"]
struct ProposeParser;

#[inline]
#[allow(clippy::result_large_err)]
pub fn parse(src: &str) -> Result<Main> {
    let inputs = ProposeParser::parse(Rule::main, src)?;
    let input = inputs.single()?;
    ProposeParser::main(input)
}

#[pest_consume::parser]
impl ProposeParser {
    #[inline]
    fn point_id(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn common_id(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn line_2p(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b)] => Ok(Object::Line2P(a, b))
        )
    }
    #[inline]
    fn circ_3p(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b), point_id(c)] => Ok(Object::Circ3P(a, b, c))
        )
    }
    #[inline]
    fn circ_or(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), numeric(b)] => Ok(Object::CircOr(a, Box::new(b))),
        )
    }
    #[inline]
    fn circ_oa(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b)] => Ok(Object::CircOA(a, b))
        )
    }
    #[inline]
    fn circ_diam(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b)] => Ok(Object::CircDiam(a, b))
        )
    }
    #[inline]
    fn trig(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b), point_id(c)] => Ok(Object::Triangle(a, b, c))
        )
    }
    #[inline]
    fn arc(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b), point_id(c)] => Ok(Object::Arc(a, b, c))
        )
    }
    #[inline]
    fn arc_o(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b), point_id(c)] => Ok(Object::ArcO(a, b, c))
        )
    }
    #[inline]
    fn polygon(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [point_id(p)..] => Ok(Object::Polygon(p.collect()))
        )
    }
    #[inline]
    fn linear(input: Node) -> Result<Linear> {
        match_nodes!(
            input.into_children();
            [common_id(a)] => Ok(Linear::Name(a)),
            [line_2p(a)] => {
                if let Object::Line2P(l, k) = a {
                    Ok(Linear::Line2P(l, k))
                } else {
                    unreachable!()
                }
            }
        )
    }
    #[inline]
    fn distance(input: Node) -> Result<Numeric> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b)] => Ok(Numeric::Distance2P(a, b)),
            [point_id(a), linear(b)] => Ok(Numeric::DistancePL(a, b)),
            [linear(a), linear(b)] => Ok(Numeric::Distance2L(a, b)),
        )
    }
    #[inline]
    fn angle_3p(input: Node) -> Result<(String, String, String)> {
        match_nodes!(
            input.into_children();
            [point_id(a), point_id(b), point_id(c)] => Ok((a, b, c))
        )
    }
    #[inline]
    fn angle_2l(input: Node) -> Result<Numeric> {
        match_nodes!(
            input.into_children();
            [linear(a), linear(b)] => Ok(Numeric::Angle2L(a, b))
        )
    }
    #[inline]
    fn number(input: Node) -> Result<f64> {
        input
            .as_str()
            .to_string()
            .parse::<f64>()
            .map_err(|e| input.error(e))
    }
    #[inline]
    fn degree(input: Node) -> Result<f64> {
        let mut x = input.as_str().to_string();
        x.pop();
        x.pop();
        x.pop();
        let x = x.parse::<f64>().map_err(|e| input.error(e))?;
        Ok(x * PI / 180.0)
    }
    #[inline]
    fn rich_number(input: Node) -> Result<f64> {
        match_nodes!(
            input.into_children();
            [number(a)] => Ok(a),
            [degree(a)] => Ok(a),
        )
    }
    #[inline]
    fn numeric(input: Node) -> Result<Numeric> {
        match_nodes!(
            input.into_children();
            [distance(a)] => Ok(a),
            [angle_3p((a, b, c))] => Ok(Numeric::Angle3P(a, b, c)),
            [angle_2l(a)] => Ok(a),
            [rich_number(a)] => Ok(Numeric::Number(a)),
            [common_id(a)] => Ok(Numeric::Name(a)),
        )
    }
    #[inline]
    fn boolean(input: Node) -> Result<bool> {
        match input.as_str() {
            "true" => Ok(true),
            "false" => Ok(false),
            _ => Err(input.error("Not a bool")),
        }
    }
    #[inline]
    fn str_inner(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn string(input: Node) -> Result<String> {
        Self::str_inner(input.into_children().next().unwrap())
    }
    #[inline]
    fn ortho_coord(input: Node) -> Result<DeclRight> {
        match_nodes!(
            input.into_children();
            [numeric(a), numeric(b)] => Ok(DeclRight::OrthoCoord(Box::new(a), Box::new(b)))
        )
    }
    #[inline]
    fn polar_coord(input: Node) -> Result<DeclRight> {
        match_nodes!(
            input.into_children();
            [numeric(a), numeric(b)] => Ok(DeclRight::PolarCoord(Box::new(a), Box::new(b)))
        )
    }
    #[inline]
    fn coord(input: Node) -> Result<DeclRight> {
        match_nodes!(
            input.into_children();
            [ortho_coord(a)] => Ok(a),
            [polar_coord(b)] => Ok(b),
        )
    }
    #[inline]
    fn eval(input: Node) -> Result<Object> {
        let string = input
            .as_str()
            .strip_prefix('$')
            .unwrap()
            .strip_suffix('$')
            .unwrap();
        Ok(Object::Eval(string.to_string()))
    }
    #[inline]
    fn any_id(input: Node) -> Result<String> {
        match_nodes!(
            input.into_children();
            [point_id(a)] => Ok(a),
            [common_id(a)] => Ok(a),
        )
    }
    #[inline]
    fn common_obj(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [line_2p(a)] => Ok(a),
            [circ_3p(a)] => Ok(a),
            [circ_or(a)] => Ok(a),
            [circ_oa(a)] => Ok(a),
            [circ_diam(a)] => Ok(a),
            [any_id(a)] => Ok(Object::Name(a))
        )
    }
    #[inline]
    fn arg(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [trig(a)] => Ok(a),
            [common_obj(a)] => Ok(a),
            [numeric(a)] => Ok(Object::Numeric(Box::new(a))),
            [eval(a)] => Ok(a),
        )
    }
    #[inline]
    fn draw_obj(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [polygon(a)] => Ok(a),
            [arc(a)] => Ok(a),
            [arc_o(a)] => Ok(a),
            [angle_3p((a, b, c))] => Ok(Object::Angle3P(a, b, c)),
            [common_obj(a)] => Ok(a),
        )
    }
    #[inline]
    fn decor_obj(input: Node) -> Result<Object> {
        match_nodes!(
            input.into_children();
            [polygon(a)] => Ok(a),
            [angle_3p((a, b, c))] => Ok(Object::Angle3P(a, b, c)),
            [arc(a)] => Ok(a),
            [arc_o(a)] => Ok(a),
            [common_obj(a)] => Ok(a),
        )
    }
    #[inline]
    fn method(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn args(input: Node) -> Result<Vec<Object>> {
        input.into_children().map(Self::arg).collect()
    }
    #[inline]
    fn expr(input: Node) -> Result<DeclRight> {
        match_nodes!(
            input.into_children();
            [method(a), args(b)] => Ok(DeclRight::Expr(a, b))
        )
    }
    #[inline]
    fn decl_right(input: Node) -> Result<DeclRight> {
        match_nodes!(
            input.into_children();
            [coord(a)] => Ok(a),
            [expr(a)] => Ok(a),
            [arg(a)] => Ok(DeclRight::Object(Box::new(a))),
            [eval(a)] => Ok(DeclRight::Object(Box::new(a))),
        )
    }
    #[inline]
    fn direct(input: Node) -> Result<DeclLeft> {
        match_nodes!(
            input.into_children();
            [any_id(a)] => Ok(DeclLeft::Direct(a))
        )
    }
    #[inline]
    fn destruct(input: Node) -> Result<DeclLeft> {
        match_nodes!(
            input.into_children();
            [any_id(a), any_id(b)] => Ok(DeclLeft::Destruct(a, b))
        )
    }
    #[inline]
    fn decl_left(input: Node) -> Result<DeclLeft> {
        match_nodes!(
            input.into_children();
            [direct(a)] => Ok(a),
            [destruct(a)] => Ok(a),
        )
    }
    #[inline]
    fn decl(input: Node) -> Result<Decl> {
        match_nodes!(
            input.into_children();
            [decl_left(a), decl_right(b)] => Ok(Decl(a, b))
        )
    }
    #[inline]
    fn config_value(input: Node) -> Result<ConfigValue> {
        match_nodes!(
            input.into_children();
            [rich_number(a)] => Ok(ConfigValue::Number(a)),
            [boolean(a)] => Ok(ConfigValue::Bool(a)),
            [string(a)] => Ok(ConfigValue::String(a)),
        )
    }
    #[inline]
    fn config_name(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn config(input: Node) -> Result<(String, ConfigValue)> {
        match_nodes!(
            input.into_children();
            [config_name(key), config_value(val)] => Ok((key, val))
        )
    }
    #[inline]
    fn configs(input: Node) -> Result<Config> {
        let items = input.into_children().flat_map(Self::config);
        let mut map = Config::new();
        for (key, val) in items {
            map.insert(key, val);
        }
        Ok(map)
    }
    #[inline]
    fn config_line(input: Node) -> Result<Config> {
        Self::configs(input.into_children().single().unwrap())
    }
    #[inline]
    fn draw_step(input: Node) -> Result<StyledObject> {
        match_nodes!(
            input.into_children();
            [draw_obj(obj), configs(config)] => Ok(StyledObject { obj, config: Some(config) }),
            [draw_obj(obj)] => Ok(StyledObject { obj, config: None })
        )
    }
    #[inline]
    fn draw(input: Node) -> Result<Draw> {
        input.into_children().map(Self::draw_step).collect()
    }
    #[inline]
    fn decoration(input: Node) -> Result<String> {
        Ok(input.as_str().to_string())
    }
    #[inline]
    fn decor_step(input: Node) -> Result<DecorObject> {
        match_nodes!(
            input.into_children();
            [decor_obj(obj), decoration(decor), configs(config)] => Ok(DecorObject { obj, decor, config: Some(config) }),
            [decor_obj(obj), decoration(decor)] => Ok(DecorObject { obj, decor, config: None })
        )
    }
    #[inline]
    fn decor(input: Node) -> Result<Decor> {
        input.into_children().map(Self::decor_step).collect()
    }
    #[inline]
    fn file_line(input: Node) -> Result<FileLine> {
        match_nodes!(
            input.into_children();
            [config_line(a)] => Ok(FileLine::Config(a)),
            [draw(a)] => Ok(FileLine::Draw(a)),
            [decor(a)] => Ok(FileLine::Decor(a)),
            [decl(a)] => Ok(FileLine::Decl(Box::new(a))),
        )
    }
    #[inline]
    fn main(input: Node) -> Result<Main> {
        input.into_children().map(Self::file_line).collect()
    }
}
