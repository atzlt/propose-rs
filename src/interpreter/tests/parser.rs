use crate::interpreter::parser::parse;

const TEST_STR: [&str; 13] = [
    "config width=14, height=10, dotsize=7, dotstroke=\"red\", dotwidth=1;",
    "O2 = (2, 0);",
    "d = @(O2, 4);",
    "A, B = i c, d;",
    "l = l 0.1, 1, A;",
    "S, _ = i l, c, A;",
    "k = l -0.2, 1, B;",
    "draw c, d, A, B, S, T, ST, P, Q, PQ;",
    "decor SP: >, TQ: >;",
    "A = (3 : -40deg);",
    "draw c[label=\"c\",loc=160deg,angle=160deg], A[label=\"A\",angle=-60deg], B[label=\"B\",angle=50deg];",
    "draw CS[color=\"blue\",dash=5], BS, DS;",
    "draw A-B-C-D[fill=\"#ff000033\"];"
];

#[test]
fn test_parser() {
    insta::with_settings!({sort_maps => true}, {
        for string in TEST_STR {
            insta::assert_yaml_snapshot!(parse(string).unwrap());
        }
    })
}
