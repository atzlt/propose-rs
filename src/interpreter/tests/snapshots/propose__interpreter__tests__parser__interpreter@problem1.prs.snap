---
source: src/interpreter/tests/parser.rs
expression: parse(&input).unwrap()
input_file: test_input/problem1.prs
---
[
  Decl(Decl(Direct("O"), OrthoCoord(Number(0.0), Number(0.0)))),
  Decl(Decl(Direct("r"), Object(Numeric(Number(4.0))))),
  Decl(Decl(Direct("c"), Object(CircOr("O", Name("r"))))),
  Decl(Decl(Direct("C"), PolarCoord(Name("r"), Number(-0.3490658503988659)))),
  Decl(Decl(Direct("A"), PolarCoord(Name("r"), Number(1.7453292519943295)))),
  Decl(Decl(Direct("B"), PolarCoord(Name("r"), Number(3.490658503988659)))),
  Decl(Decl(Direct("D"), Expr("i", [
    Line2P("A", "O"),
    Name("c"),
    Name("A"),
  ]))),
  Decl(Decl(Direct("x"), Object(Numeric(Angle3P("B", "A", "C"))))),
  Decl(Decl(Direct("x"), Object(Eval(" -180 * deg + 2 * x ")))),
  Decl(Decl(Direct("E"), Expr("on", [
    Name("A"),
    Name("B"),
    Numeric(Number(0.7)),
  ]))),
  Decl(Decl(Direct("F"), Expr("rot", [
    Name("E"),
    Name("D"),
    Name("x"),
  ]))),
  Decl(Decl(Direct("F"), Expr("i", [
    Line2P("D", "F"),
    Line2P("A", "C"),
  ]))),
  Decl(Decl(Direct("l"), Expr("ab", [
    Name("E"),
    Name("D"),
    Name("F"),
  ]))),
  Decl(Decl(Direct("G"), Expr("i", [
    Name("l"),
    Line2P("E", "F"),
  ]))),
  Draw([
    StyledObject(
      obj: Name("c"),
      config: None,
    ),
    StyledObject(
      obj: Name("A"),
      config: None,
    ),
    StyledObject(
      obj: Name("B"),
      config: None,
    ),
    StyledObject(
      obj: Name("C"),
      config: None,
    ),
    StyledObject(
      obj: Name("D"),
      config: None,
    ),
    StyledObject(
      obj: Name("E"),
      config: None,
    ),
    StyledObject(
      obj: Name("F"),
      config: None,
    ),
    StyledObject(
      obj: Name("G"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("A", "B"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("B", "C"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("C", "A"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("A", "D"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("D", "E"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("D", "F"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("E", "F"),
      config: None,
    ),
    StyledObject(
      obj: Line2P("D", "G"),
      config: None,
    ),
  ]),
  Config({
    "decorcolor": String("blue"),
  }),
  Decl(Decl(Direct("K"), Expr("i", [
    Line2P("D", "G"),
    Line2P("B", "C"),
  ]))),
  Decor([
    DecorObject(
      obj: Line2P("D", "K"),
      decor: "|",
      config: None,
    ),
    DecorObject(
      obj: Line2P("K", "G"),
      decor: "|",
      config: None,
    ),
  ]),
]
